mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp

```diff
@@ -1446,6 +1446,117 @@ struct ChainedReduction final : OpRewritePattern<vector::ReductionOp> {
+///    %res = arith.addi %0, %arg2 : vector<8xi32>
+/// ```
+/// ATM this pattern is limited to `vector.shape_cast` ops that fold the unit
+/// dim, e.g.:
+/// ```
+///   vector.shape_cast %mul : vector<1x4xf32> to vector<4xf32>
+/// ```
+/// In addition, the input vector should be the result of an arithmetic
+/// operation, `ArithOp`.
+template <typename ArithOp>
+struct ReorderShapeCastWithUnitDimAndArith
+    : public OpRewritePattern<vector::ShapeCastOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ShapeCastOp shapeCastOp,
+                                PatternRewriter &rewriter) const override {
+    auto arithOp = shapeCastOp.getSource().getDefiningOp<ArithOp>();
+    if (!arithOp)
+      return failure();
+
+    // All arith ops are elementwise - filter out everything else.
+    if (!arithOp.template hasTrait<OpTrait::Elementwise>())
+      return failure();
+
+    // TODO: Add support for unary ops
+    if (arithOp->getOperands().size() != 2)
+      return failure();
+
+    auto sourceVectorType =
+        dyn_cast<VectorType>(shapeCastOp.getSource().getType());
+    auto resultVectorType =
+        dyn_cast<VectorType>(shapeCastOp.getResult().getType());
+    if (!sourceVectorType || !resultVectorType)
+      return failure();
+
+    // Either the leading or the trailing dims of the input should be
+    // non-scalable 1.
+    bool leadDimUnitFixed = ((sourceVectorType.getShape().back() != 1) ||
+                             (sourceVectorType.getScalableDims().back()));
+    bool trailinDimUnitFixed = ((sourceVectorType.getShape().front() != 1) ||
+                                (sourceVectorType.getScalableDims().front()));
+    if (!leadDimUnitFixed && !trailinDimUnitFixed)
+      return failure();
+
+    // Does this shape_cast fold the input vector?
+    if (resultVectorType.getRank() != (sourceVectorType.getRank() - 1))
+      return failure();
+
+    // Does this shape_cast fold the traling/leading _unit_ dim?
+    // TODO: Even when the trailing/leading unit dims are folded, there might
+    // still be some "inner" unit dims left.
+    if (llvm::any_of(resultVectorType.getShape(),
+                     [](int64_t dim) { return (dim == 1); }))
+      return failure();
+
+    auto loc = shapeCastOp->getLoc();
+
+    // shape_cast(a)
+    auto lhs = rewriter.create<vector::ShapeCastOp>(loc, resultVectorType,
+                                                    arithOp->getOperands()[0],
+                                                    shapeCastOp->getAttrs());
+    // shape_cast(b)
+    auto rhs = rewriter.create<vector::ShapeCastOp>(loc, resultVectorType,
+                                                    arithOp->getOperands()[1],
+                                                    shapeCastOp->getAttrs());
+
+    // Replace shape_cast(a ArithOp b) with shape_cast(a) ArithOp shape_cast(b)
+    rewriter.replaceOpWithNewOp<ArithOp>(shapeCastOp, lhs, rhs);
```

Comment:
To replace a new op in a generic way we can use the OperationState approach and extract all the op fields from the original operation.

Reply:
I'm not sure what you mean by this - would you have an example?

Reply:
https://github.com/llvm/llvm-project/blob/8eff5704829ba5edd28754fd9ec7665b34fde22a/mlir/lib/Dialect/Utils/StructuredOpsUtils.cpp#L200-L204

That's a way to create an operation without actually knowing which operation you are creating (and without using a template) :)

Reply:
OK, I see what you meant, thanks!

I'm actually doing something similar already (see [here](https://github.com/llvm/llvm-project/pull/74817/files#diff-f580c418633c6029c99785e127b040089cef3a6b94006a2495ae38e3a2ca747dR1515-R1517)). Not using OperationState directly, but also not requiring template param :) (not sure using OperationState directly would be better, but I'm always open to suggestions)